voicevox_engine/tts_pipeline/tts_engine.py

```diff
@@ -232,6 +235,55 @@ def raw_wave_to_output_wave(
     return wave
 
 
+def _hira_to_kana(text: str) -> str:
+    """ひらがなをカタカナに変換する"""
+    return "".join(chr(ord(c) + 96) if "ぁ" <= c <= "ゔ" else c for c in text)
+
+
+def calc_phoneme_lengths(
+    consonant_lengths: NDArray[np.int64],
+    note_durations: NDArray[np.int64],
+) -> NDArray[np.int64]:
+    """
+    子音長と音符長から音素長を計算する
+    ただし、母音はノートの頭にくるようにするため、
+    予測された子音長は前のノートの長さを超えないように調整される
+    """
+    phoneme_durations = []
+    for i in range(len(consonant_lengths)):
+        if i < len(consonant_lengths) - 1:
+            # 最初のノートは子音長が0の、pauである必要がある
+            if i == 0 and consonant_lengths[i] != 0:
+                raise HTTPException(
+                    status_code=400,
+                    detail=f"consonant_lengths[0] must be 0, but {consonant_lengths[0]}",
+                )
+
+            next_consonant_length = consonant_lengths[i + 1]
+            note_duration = note_durations[i]
+
+            # もし、次のノートの子音長が負になる場合、現在のノートの半分にする
+            if next_consonant_length < 0:
+                next_consonant_length = consonant_lengths[i + 1] = note_duration // 2
```

Comment:
ここ、普通に考えると１フレームにするとかのが良い気がしました。
（というか非負はコアの方に書くべきかも･･･？）

Reply:
普通に考えれば1にするのが正しそうですが、子音長を1にすると、滑舌が怪しくなることがあるので、できれば避けたいかもです。
とはいえ、前のノートが長い場合、子音長もつられて長くなりそうなので、少し検討した方がいいなとは思いますね...

とはいえ負になるのはまずモデルが悪いので、コア側で調整するのが正しいそうですね...

Reply:
> 普通に考えれば1にするのが正しそうですが、子音長を1にすると、滑舌が怪しくなることがあるので、できれば避けたいかもです。

気持ちはわかるのですが、それはそれで1だと推論したものがそのままになるので、意図と実装がずれてる雰囲気を感じます。
触ってないのでわからないのですが、子音長が1だった場合は2にする、とかでも良いきがしてます。

とりあえずマージを目指すのが良さそうな気がするので、非負になる処理をコアに足すことをFIXMEコメントにしてとりあえず後回しにするのはどうでしょう。
```suggestion
            # もし、次のノートの子音長が負になる場合、現在のノートの半分にする
            # FIXME: 非負にならないようコア側に処理を足す
            if next_consonant_length < 0:
                next_consonant_length = consonant_lengths[i + 1] = note_duration // 2
```